rethinkdb/datadog_checks/rethinkdb/metrics/statistics.py

```diff
@@ -0,0 +1,274 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import logging
+from typing import Iterator
+
+from ..connections import Connection
+from ..queries import QueryEngine
+from ..types import Metric
+
+logger = logging.getLogger(__name__)
+
+
+def collect_cluster_statistics(engine, conn):
+    # type: (QueryEngine, Connection) -> Iterator[Metric]
+    """
+    Collect metrics about cluster statistics.
+
+    See: https://rethinkdb.com/docs/system-stats#cluster
+    """
+    logger.debug('collect_cluster_statistics')
+
+    stats = engine.query_cluster_stats(conn)
+    logger.debug('cluster_statistics stats=%r', stats)
+
+    query_engine = stats['query_engine']
+
+    yield {
+        'type': 'gauge',
+        'name': 'rethinkdb.stats.cluster.queries_per_sec',
+        'value': query_engine['queries_per_sec'],
+        'tags': [],
+    }
+
+    yield {
+        'type': 'gauge',
+        'name': 'rethinkdb.stats.cluster.read_docs_per_sec',
+        'value': query_engine['read_docs_per_sec'],
+        'tags': [],
+    }
+
+    yield {
+        'type': 'gauge',
+        'name': 'rethinkdb.stats.cluster.written_docs_per_sec',
+        'value': query_engine['written_docs_per_sec'],
+        'tags': [],
+    }
+
+
+def collect_server_statistics(engine, conn):
+    # type: (QueryEngine, Connection) -> Iterator[Metric]
+    """
+    Collect metrics about server statistics.
+
+    See: https://rethinkdb.com/docs/system-stats#server
+    """
+    logger.debug('collect_server_statistics')
+
+    for server, stats in engine.query_servers_with_stats(conn):
+        logger.debug('server_statistics server=%r stats=%r', server, stats)
+
+        name = server['name']
+        server_tags = server['tags']
+        query_engine = stats['query_engine']
+
+        tags = ['server:{}'.format(name)] + server_tags
+
+        yield {
+            'type': 'gauge',
+            'name': 'rethinkdb.stats.server.client_connections',
+            'value': query_engine['client_connections'],
```

Comment:
I _could_ factor out…

But currently this simple approach fully benefits from type checking as a fool-proofing strategy. Eg all the literals passed to `query_engine[...]` are type-checked, same for `type`, etc.

This approach has a "the simplest, dumbest code one can ever write" feel to it that makes it pretty easy to maintain IMO. (OTOH a list constant would already add an extra level of indirection…

(Side note: with the suggested snippet, I'd actually have to ditch my `TypedDict` or silence mypy because it'd complain about the dynamic key access.)

Thoughts?

Reply:
I'm not that familiar yet with all the typing features, but there is maybe a way to have both ? typing and factored code?

Repeating a block of code X times seems not right.

Maintainability wise, with factored code if you add a new metrics, it's a 1 line chance, instead of 6+ lines for the non factored way.

Reply:
I agree with Alex here. Maintainability should trump type hinting in all cases

Reply:
I don't agree that more code always means less maintainable code — in particular in this case I think it's quite the opposite.

I'll do the work and factor things out to see how it looks and we can compare, but almost certainly these functions will end up using ad-hoc, abstract logic that'll be hard to understand and change further down the line (e.g. me in 6 months).

Sure, the current code is verbose, but:
- It's plain, simple, explicit, dead-simple to change.
- As for silly typos on copy-pastes during maintenance: type checking has our back (it will catch them! because all literals here are type-checked). Type checking is working _for_ us, not _against_ us, here.

At least that's my view here, and I know everyone's mileage may vary. But as I said, I'll factor out and we'll see, at least it's interesting to compare both options. :-)

Reply:
Maintainability is a grey area. But in this case it seems you're making things gratuitously different. We have 150 integrations, I think we can try to find inspirations from existing patterns. In most cases, we defined static mappings of stats -> metrics, and i think it served us well.

At the bare minimum, you shouldn't need to mention `rethinkdb.stats.server` 8 times or so.

Reply:
Okay, I'm sold on the idea; working on a refactor to make this more declarative. :-)

Reply:
I pushed a "let's factor things out" refactor in e60c27d (still needs tackling `config.py` and `current_issues.py`, but the bulk of the work is there).

Note this hasn't changed the LOC count (+3060 before, +3090 after), contrary to what I initially expected.

I'm actually not entirely convinced yet this will facilitate maintenance a lot (maybe the contrary, because as I said we now have to deal with a kind-of-a-beast ad-hoc abstraction), but my implementation/API may not be as tight as it could yet.

I'll refine tomorrow, but let me know what you think 😄Could be the beginning of a `DocumentQueryManager`…

Reply:
Moved draft to https://github.com/DataDog/integrations-core/pull/6092 so I can keep iterating/fixing things here without conflicting with the refactoring.

Reply:
Refactored in 130fc74e513c900f2a50fd3ac4d914cf7e5becf6